bugsnag-android-core/src/main/java/com/bugsnag/android/internal/DocumentPath.kt

```diff
@@ -0,0 +1,209 @@
+ * mechanism using the backslash character:
+ *
+ * \. = literal dot
+ * \\ = literal backslash
+ *
+ * Path components are either integers or non-integers. Integers refer to the previous path
+ * component as a list, providing the index to look up. Non-integers refer to the previous path
+ * component as a map, providing the key to look up (for example, in the path "a.b.3", "a" is
+ * treated as a map and "b" is treated as a list, so the path refers to document root with map
+ * key "a", which is then indexed with map key "b", which is then indexed with list index 3).
+ *
+ * Any nonexistent parts of the document are created on-demand as they are encountered while
+ * resolving paths (either as a list in the case of integer path components, or as a map for
+ * non-integers). In the case of lists, only index 0 of a nonexistent list is valid (in which
+ * case the list will be created, and the value stored at index 0).
+ *
+ * As well, the following special rules apply:
+ *
+ * - An empty path component refers to the document root, and causes the value to replace
+ * the entire document.
+ * - The integer path component -1 refers to the last index in the array, or index 0 if the
+ * array is empty or nonexistent.
+ * - If a path ends in a dot (e.g. events.exceptions.stacktrace.), It means that the last
+ * path entry (e.g. stacktrace) is to be treated as a list, and the value is to be appended
+ * to that list.
+ *
+ * The document to be modified must be serializable into JSON. The following types are supported:
+ * - null
+ * - boolean
+ * - integer (max 15 digits)
+ * - float
+ * - string
+ * - List(Object)
+ * - Map(String, Object)
+ */
+class DocumentPath(path: String) {
+    val directives: List<DocumentPathDirective> = toPathDirectives(path)
+
+    /**
+     * Modify a document, setting this path in the document to the specified value.
+     *
+     * @param document The document to modify.
+     * @param value    The value to modify with.
+     * @return The modified document (may not be the same document that was passed in).
+     */
+    fun modifyDocument(document: MutableMap<String, Any>, value: Any?): MutableMap<String, Any> {
+        if (directives.isEmpty()) {
+            requireNotNull(value) { "Cannot replace entire document with null" }
+            require(value is MutableMap<*, *>) { "Value replacing document must be a mutable map" }
+            @Suppress("UNCHECKED_CAST")
+            return value as MutableMap<String, Any>
+        }
+        val updatedDocument = fillInMissingContainers(document, 0)!!
+        applyValueToContainer(value, updatedDocument)
+        @Suppress("UNCHECKED_CAST")
+        return updatedDocument as MutableMap<String, Any>
+    }
+
+    // Fill in the document parts that will be navigated, but don't exist yet.
+    private fun fillInMissingContainers(parent: Any?, index: Int): Any? {
+        var updatedParent = parent
```

Comment:
If `parent` is null, `updatedParent` gets assigned a new container after the directive is extracted. Otherwise `updatedParent` remains an alias to `parent`.

This is to build any container levels that aren't already present in the document we're modifying.